sa/sa.go

```diff
@@ -332,15 +340,50 @@ func (t TooManyCertificatesError) Error() string {
 // The highest count this function can return is 10,000. If there are more
 // certificates than that matching one of the provided domain names, it will return
 // TooManyCertificatesError.
+// Queries will be run in parallel. If any of them error, only one error will
+// be returned.
 func (ssa *SQLStorageAuthority) CountCertificatesByNames(ctx context.Context, domains []string, earliest, latest time.Time) ([]*sapb.CountByNames_MapElement, error) {
-	var ret []*sapb.CountByNames_MapElement
+	work := make(chan string, len(domains))
+	type result struct {
+		err    error
+		count  int
+		domain string
+	}
+	results := make(chan result, len(domains))
 	for _, domain := range domains {
-		currentCount, err := ssa.countCertificatesByName(domain, earliest, latest)
-		if err != nil {
-			return ret, err
+		work <- domain
+	}
+	close(work)
+	var wg sync.WaitGroup
+	// We may perform up to 100 queries, depending on what's in the certificate
+	// request. Parallelize them so we don't hit our timeout, but limit the
+	// parallelism so we don't consume too many threads on the database.
+	for i := 0; i < ssa.parallelismPerRPC; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for domain := range work {
+				currentCount, err := ssa.countCertificatesByName(domain, earliest, latest)
+				if err != nil {
+					results <- result{err: err}
+					continue
```

Comment:
kind-of-a-nit: why `continue`? If a single error is returned we fail out of this method so it seems like once that happens we probably shouldn't do any more work just for the sake of saving resources.

Perhaps a simple method of signalling to the other goroutines to just return (i.e. a signal `chan` or something)?